src/libtest/formatters/json.rs

```diff
@@ -0,0 +1,129 @@
+
+use ::*;
+
+pub(crate) struct JsonFormatter<T> {
+    out: OutputLocation<T>,
+}
+
+impl<T: Write> JsonFormatter<T> {
+    pub fn new(out: OutputLocation<T>) -> Self {
+        Self { out }
+    }
+
+    fn write_message(&mut self, s: &Json) -> io::Result<()> {
+        serde_json::to_writer(&mut self.out, s)
+                        .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+        self.out.write_all(b"\n")
+    }
+}
+
+impl<T: Write> OutputFormatter for JsonFormatter<T> {
+    fn write_run_start(&mut self, test_count: usize) -> io::Result<()> {
+        self.write_message(&json!({
+            "type": "suite",
+            "event": "started",
+            "test_count": test_count,
+        }))
+    }
+
+    fn write_test_start(&mut self, desc: &TestDesc) -> io::Result<()> {
+        self.write_message(&json!({
+            "type": "test",
+            "event": "started",
+            "name": desc.name.as_str(),
+        }))
+    }
+
+    fn write_result(
+        &mut self,
+        desc: &TestDesc,
+        result: &TestResult,
+        stdout: &[u8],
+    ) -> io::Result<()> {
+        match *result {
+            TrOk => self.write_message(&json!({
+                "type": "test",
+                "name": desc.name.as_str(),
+                "event": "ok",
+            })),
+            TrFailed => self.write_message(&json!({
+                "type": "test",
+                "name": desc.name.as_str(),
+                "event": "failed",
+                "stdout": String::from_utf8_lossy(stdout),
+            })),
+            TrFailedMsg(ref m) => self.write_message(&json!({
+                "type": "test",
+                "name": desc.name.as_str(),
+                "event": "failed",
+                "message": m,
+            })),
+            TrIgnored => self.write_message(&json!({
+                "type": "test",
+                "name": desc.name.as_str(),
+                "event": "ignored",
+            })),
+            TrAllowedFail => self.write_message(&json!({
+                "type": "test",
+                "name": desc.name.as_str(),
+                "event": "allowed_failure",
+            })),
+            TrBench(ref bs) => {
+                let median = bs.ns_iter_summ.median as usize;
+                let deviation = (bs.ns_iter_summ.max - bs.ns_iter_summ.min) as usize;
+
+                if bs.mb_s == 0 {
+                    self.write_message(&json!({
+                        "type": "bench",
+                        "name": desc.name.as_str(),
+                        "median": median,
+                        "deviation": deviation,
+                    }))
+                } else {
+                    self.write_message(&json!({
+                        "type": "bench",
+                        "name": desc.name.as_str(),
+                        "median": median,
+                        "deviation": deviation,
+                        "mib_per_second": bs.mb_s,
```

Comment:
Maybe collapse the if and always output `mib_per_second` but with `null` if there is no data:

```rust
"mib_per_second": if bs.mb_s == 0 { None } else { Some(bs.mb_s) },
```